src/Service/Services/ExecutionHelper.cs

```diff
@@ -0,0 +1,431 @@
+                            argumentSchema: argument,
+                            variables: variables));
+                }
+            }
+
+            // Overwrite the default values with the passed in arguments
+            IReadOnlyList<ArgumentNode> passedArguments = query.Arguments;
+
+            foreach (ArgumentNode argument in passedArguments)
+            {
+                string argumentName = argument.Name.Value;
+                IInputField argumentSchema = argumentSchemas[argumentName];
+
+                if (parameters.ContainsKey(argumentName))
+                {
+                    parameters[argumentName] =
+                        ExtractValueFromIValueNode(
+                            value: argument.Value,
+                            argumentSchema: argumentSchema,
+                            variables: variables);
+                }
+                else
+                {
+                    parameters.Add(
+                        argumentName,
+                        ExtractValueFromIValueNode(
+                            value: argument.Value,
+                            argumentSchema: argumentSchema,
+                            variables: variables));
+                }
+            }
+
+            return parameters;
+        }
+
+        /// <summary>
+        /// InnerMostType is innermost type of the passed Graph QL type.
+        /// This strips all modifiers, such as List and Non-Null.
+        /// So the following GraphQL types would all have the underlyingType Book:
+        /// - Book
+        /// - [Book]
+        /// - Book!
+        /// - [Book]!
+        /// - [Book!]!
+        /// </summary>
+        internal static IType InnerMostType(IType type)
+        {
+            if (type.ToString() == type.InnerType().ToString())
+            {
+                return type;
+            }
+
+            return InnerMostType(type.InnerType());
+        }
+
+        public static InputObjectType InputObjectTypeFromIInputField(IInputField field)
+        {
+            return (InputObjectType)(InnerMostType(field.Type));
+        }
+
+        private static IDictionary<string, object?> GetParametersFromContext(
+            IMiddlewareContext context)
+        {
+            return GetParametersFromSchemaAndQueryFields(
+                context.Selection.Field,
+                context.Selection.SyntaxNode,
+                context.Variables);
+        }
+
+        /// <summary>
+        /// Get metadata from context
+        /// </summary>
+        private static IMetadata GetMetadata(IPureResolverContext context)
+        {
+            // The pure resolver context has no access to the scoped context data,
+            // I will change that for version 14. The following code is a workaround
+            // and stores the metadata per root field on the global context.
+            return (IMetadata)context.ContextData[GetMetadataKey(context.Path)]!;
```

Comment:
> stores the metadata per root field on the global context.

There are going to be many metadata objects for a given query. The way this is written, we'll get many duplicate keys which results in exceptions. And if we overwrite the metadata object stored using Key: RootObjectTypeName Value: Metadata  every time we execute `ResolveObject` and `ResolveList` we lose track of metadata.

```graphql
query filterparamListQuery {
  books(filter: { id: { gte: 1 }, and: [{ id: { lte: 4 } }] }) {
    items {
      id
      publishers {
        books(first: 3, filter: { id: { neq: 2 } }) {
          items {
            id
          }
        }
      }
    }
  }
}
```
Will have metadata (PaginationMetadata) objects for each level:
- Level 1: `books` with metadata.subqueries containing "items" key
- Level 2: `items` with metadata.subqueries containing "publishers" key
- Level 3: `publishers` with metadata.subqueries containing "items" key
- Level 4: `items` with metadata.subqueries containing "books" key
- Level : `books` no additional subqueries from here.

Reply:
We could also use the scoped context data which is branched for each level. Is basically an immutable dictionary.

Reply:
context.ScopedContextData

Reply:
@michaelstaib, Maybe I misunderstand, but your original comment says: "The pure resolver context has no access to the scoped context data." Is that not the case anymore?

Reply:
because we are still using HC12

Reply:
Yes .... sorry, In 13 we have access to this now :)